core/services/balance_monitor.go

```diff
@@ -0,0 +1,141 @@
+package services
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/core/assets"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+	"github.com/smartcontractkit/chainlink/core/store"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+
+	gethCommon "github.com/ethereum/go-ethereum/common"
+	"github.com/pkg/errors"
+)
+
+// BalanceMonitor checks the balance for each key on every new head
+type BalanceMonitor interface {
+	store.HeadTrackable
+	GetEthBalance(gethCommon.Address) *assets.Eth
+}
+
+type balanceMonitor struct {
+	store             *store.Store
+	gethClientWrapper store.GethClientWrapper
+	ethBalances       map[gethCommon.Address]*assets.Eth
+	ethBalancesMtx    *sync.RWMutex
+}
+
+// NewBalanceMonitor returns a new balanceMonitor
+func NewBalanceMonitor(store *store.Store, gethClientWrapper store.GethClientWrapper) BalanceMonitor {
+	return &balanceMonitor{
+		store:             store,
+		gethClientWrapper: gethClientWrapper,
+		ethBalances:       make(map[gethCommon.Address]*assets.Eth),
+		ethBalancesMtx:    new(sync.RWMutex),
+	}
+}
+
+// Connect complies with HeadTrackable
+func (bm *balanceMonitor) Connect(_ *models.Head) error {
+	// Connect head can be out of date, so always query the latest balance
+	bm.checkBalance(nil)
+	return nil
+}
+
+// Disconnect complies with HeadTrackable
+func (bm *balanceMonitor) Disconnect() {}
+
+const ethFetchTimeout = 2 * time.Second
+
+// OnNewLongestChain checks the balance for each key
+func (bm *balanceMonitor) OnNewLongestChain(head models.Head) {
+	bm.checkBalance(&head)
+}
+
+func (bm *balanceMonitor) checkBalance(head *models.Head) {
+	keys, err := bm.store.Keys()
+	if err != nil {
+		logger.Error("BalanceMonitor: error getting keys", err)
+	}
+
+	var wg sync.WaitGroup
+
+	for _, key := range keys {
+		wg.Add(1)
+
+		go func(k models.Key) {
+			ctx, cancel := context.WithTimeout(context.Background(), ethFetchTimeout)
+			defer cancel()
+
+			var headNum *big.Int
+
+			err := bm.gethClientWrapper.GethClient(func(c eth.GethClient) error {
+				var err error
+				var bal *big.Int
+				if head == nil {
+					headNum = nil
+					bal, err = c.BalanceAt(ctx, k.Address.Address(), nil)
+				} else {
+					headNum = big.NewInt(head.Number)
+					bal, err = c.BalanceAt(ctx, k.Address.Address(), headNum)
+				}
+				if err != nil {
+					return errors.Wrap(err, "error getting balance")
+				}
+				ethBal := assets.NewEth(bal.Int64())
+
+				bm.updateBalance(ethBal, k.Address.Address(), headNum)
+				return nil
+			})
+
+			if err != nil {
+				logger.Errorw(fmt.Sprintf("BalanceMonitor: error getting balance for key %s: %s", k.Address.Hex(), err.Error()), "err", err, "address", k.Address, "headNum", headNum)
+			}
+
+			wg.Done()
```

Comment:
Should we defer this call at the start of the func? Looks like we could skip it atm.

Reply:
Good catch! I had to move some error handling around to satisfy errcheck and I missed this one.

Reply:
Actually, I just double-checked and we're ok. It will always get called regardless of code path.

Note that the returns on lines 88 and 93 return from the function closure opened on line 77.

Reply:
Ah, my bad.